### PR TITLE
[BXMSPROD-670] generate CVE content for RHSA errata

### PR DESCRIPTION
**Pull request by rgdoliveira**
Issue: https://issues.redhat.com/browse/BXMSPROD-670

### CODE REVIEW

File: vars/rhsaErrata.groovy [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * Get the reference link used in the Errata.
+ * The link will always point to the higher CVE released as part of the Errata
+ *
+ * @param cveList a list of CVE
+ * @return the reference link
+ */
+def getReferenceLink(List<CVE> cveList) {
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:78]:
```suggestion
def getReferenceLink(List<CVE> cveList, String link) {
```

File: vars/rhsaErrata.groovy [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    def link = 'https://access.redhat.com/security/updates/classification/#'
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:79]: shouldn't it be env agnostic? shouldn't it be part of the jenkins global env variables? wdyt?

File: vars/rhsaErrata.groovy [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * Get the higher CVE impact.
+ *
+ * @param cveList a list of CVE
+ * @return the higher CVE impact
+ */
+def getHigherCVEImpact(List<CVE> cveList) {
+    def impactNumber = []
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:90]:
```suggestion
    def impactNumber = cveList.inject([]) {acc, cve -> 
        acc.add(CVE.impactOrdering.get(cve.impact))
        acc
    }.sort()
```
I think it should work...

File: vars/rhsaErrata.groovy [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    cveList.each {cve -> impactNumber.add(CVE.impactOrdering.get(cve.impact))}
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:91]:
```suggestion
```

File: vars/rhsaErrata.groovy [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    impactNumber.sort()
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:92]:
```suggestion
```

File: vars/rhsaErrata.groovy [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+/**
+ * Get the Bugzilla number.
+ *
+ * @param description field as part of the CVE Json
+ * @return Bugzilla number
+ */
+def getBZNumber(String description) {
+    def bzLink = description.findAll('bugzilla.redhat.com\\/show_bug.cgi\\?id=[0-9]*')[0]
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:113]: the same here about env agnostic...

File: vars/rhsaErrata.groovy [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    cveList.each {cve ->
+        def cveImpactOrdering = CVE.impactOrdering.get(cve.impact)
+        def description = (problemDescriptions.get(cveImpactOrdering) != null) ? problemDescriptions.get(cveImpactOrdering) : []
+        description.add(cve.problemDescription)
+        Collections.sort(description, String.CASE_INSENSITIVE_ORDER)
+        problemDescriptions.put(cveImpactOrdering, description)
+    }
+    problemDescriptions = problemDescriptions.sort()
+
+    def sortedDescriptions = problemDescriptions.inject([]){list, key, descriptions ->
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:161]:
```suggestion
    return problemDescriptions.inject([]){list, key, descriptions ->
```

File: vars/rhsaErrata.groovy [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        descriptions.each {list << it}
+        return list
+    }
+    return sortedDescriptions
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:165]:
```suggestion
```

File: vars/rhsaErrata.groovy [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+/**
+ * Sort problem descriptions by CVE impact and also alphabetically.
+ * This needed to be implemented this way as Jenkins Groovy CPS fails to get properly sort from a class (compareTo method)
+ *
+ * @param a list of CVE
+ * @return sorted list of problem descriptions
+ */
+def sortProblemDescriptions(List<CVE> cveList) {
+    Map<String, List<String>> problemDescriptions = [:]
```
Comment (Ginxo) [inline vars/rhsaErrata.groovy:151]:
```suggestion
    Map<String, List<String>> problemDescriptions = cveList.inject([:]) {acc, cve ->
        def cveImpactOrdering = CVE.impactOrdering.get(cve.impact)
        def description = acc.get(cveImpactOrdering) ?: [] 
        description.add(cve.problemDescription)
        Collections.sort(description, String.CASE_INSENSITIVE_ORDER)
        acc.put(cveImpactOrdering, description)
        acc
    }.sort()
```

about `def description = acc.get(cveImpactOrdering) ?: [] ` see https://groovy-lang.org/operators.html#_elvis_assignment_operator

the lines from `cveList.each ...` to `problemDescriptions = problemDescriptions.sort()` should be removed

**Review by Ginxo — CHANGES_REQUESTED** [review]
just few technical proposals. Good job @rgdoliveira

**Comment by rgdoliveira** [comment]
@Ginxo Thanks for the review, I've applied your suggestions.

**Review by Ginxo — APPROVED** [review]
LGTM

**Pull request merged**